Issue #14084: Remove checkerframework violation for DefaultConfigurat…

Rohanraj123 opened the pull request:
related #14084

nrmancuso commented:
@Rohanraj123 let's get existing PRs out of draft mode before opening more: https://github.com/checkstyle/checkstyle/pulls/Rohanraj123

nrmancuso commented on config/checker-framework-suppressions/checker-nullness-optional-interning-suppressions.xml:
```diff
@@ -140,17 +140,6 @@
-    <fileName>src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java</fileName>
-    <specifier>return</specifier>
-    <message>incompatible types in return.</message>
-    <lineContent>return children.toArray(</lineContent>
```
Annotation must be placed on the method that has this line as it's return statement. Please take some time to read through the checker framework docs to understand this tool: https://checkerframework.org/manual/

Rohanraj123 replied:
> Please take some time to read through the checker framework docs to understand this tool: https://checkerframework.org/manual/

On it sir.

Rohanraj123 commented:
> @Rohanraj123 let's get existing PRs out of draft mode before opening more: https://github.com/checkstyle/checkstyle/pulls/Rohanraj123

Not gonna open any more before these !!

Rohanraj123 commented:
Reasoning before this change : 
Heuristics for Handling Common Cases:
The Nullness Checker uses certain heuristics to handle common cases when determining the nullness type of the returned array:
If the passed array has a length of 0 or is created with size 0 (new String[0]), the returned array component type is @NonNull.
And This is the case with us too.

Reason for SuppressWarning("nullness")
Due to the complexities in analyzing toArray method calls, the explanation suggests that you may need to use @SuppressWarnings("nullness") annotation when overriding the toArray method. 

Document reference : https://checkerframework.org/manual/#null-defaults

Rohanraj123 commented:
> Reason for SuppressWarning("nullness")

Okay After using this SupressionWarnings, I saw it was not allowed. As such. But I think this is the only way to resolve this issue as per the checker framework documentation. Can you please suggest me some way via which I can use suppression warnings in DefaultConfiguration.java file.

nrmancuso commented on src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java:
```diff
@@ -103,8 +102,9 @@ public String getProperty(String propertyName) throws CheckstyleException {
     public Configuration[] getChildren() {
-        return children.toArray(
-                EMPTY_CONFIGURATION_ARRAY);
+        return Arrays.stream(children.toArray(new Configuration[0]))
```
Why are we converting a list to an array and streaming it, instead of directly streaming the array?

Rohanraj123 replied:
If not converting the list to array, yields new pitest error.

romani commented:
@Rohanraj123, do you have time to continue now?
Or we can close it for now and you will reopen it when you have time to continue it.

Rohanraj123 commented:
> @Rohanraj123, do you have time to continue now? Or we can close it for now and you will reopen it when you have time to continue it.

I will contiinue my work from here.

Rohanraj123 commented:
@romani I got the problem now. 
Basically as per the checker framework error. It says that method return type is `NonNull` whereas return expression is `Nullable`.
This is because how checker framework analyzes the `toArray`() method.
Official source : 
`The nullness semantics of [Collection.toArray(T[])](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Collection.html#toArray(T[])) cannot be captured by just the nullness type system`

returned array is of type @NonNull if : 
1 . collections type is Nonnull
2. size(passed array) <= size(collections)

Since we have children list that is converted into array which is non null because its a  initialized list of configuration
and same goes with passed array. 

Therefor as per my understanding Two possible solutions are here :
1. Use suppression warnings
2. using dynamically created array parameter. 

As I did in this commit . Can you suggest any other method if this ain't work

Rohanraj123 commented:
@romani This PR is ready for review !

Changes : 
To remove violations of checker framework. 
--> I just removed the constant predefined due to the lack of analysis of checker framework in toArray() framework. It sometimes becomes overly cautious. 
--> So I passed the dynamically created array in toArray() method to guarantee the checker framework that param array is not null.

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java:
```diff
@@ -104,7 +101,7 @@ public String getProperty(String propertyName) throws CheckstyleException {
     public Configuration[] getChildren() {
         return children.toArray(
-                EMPTY_CONFIGURATION_ARRAY);
+                new Configuration[0]);
```
This means you found defect in checkerframefork tool.
Please create issue on it, close this PR, make new PR with comments over xml suppression blocks with comment `<-- until #xxxxxx -->`

romani reviewed (changes requested):
Item

Rohanraj123 commented:
@ArrayLen() annotation can only work when Constant value checker is configured. And After looking pom.xml file, constant value checker is configured. 
so now. adding command line option `Alint=trustArrayLenZero` in compilerargs should process annotation @ArrayLen() .

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java:
```diff
@@ -36,10 +38,9 @@ public final class DefaultConfiguration implements Configuration {
 
     /** A unique serial version identifier. */
     private static final long serialVersionUID = 1157875385356127169L;
-
```
Restore this line

romani commented on src/main/java/com/puppycrawl/tools/checkstyle/DefaultConfiguration.java:
```diff
@@ -36,10 +38,9 @@ public final class DefaultConfiguration implements Configuration {
     /** Constant for optimization. */
-    private static final Configuration[] EMPTY_CONFIGURATION_ARRAY = new Configuration[0];
-
+    private static final @ArrayLen(0) Configuration[] EMPTY_CONFIGURATION_ARRAY =
```
Put annotation above, not on same line

Rohanraj123 replied:
@romani putting annotation above giving me violation in `checker-formatter` that it should be either before or after of the `private static final `

romani replied:
Please put it before

Rohanraj123 replied:
done!

Rohanraj123 replied:
> @romani putting annotation above giving me violation in `checker-formatter` that it should be before type and after modifiers.

romani reviewed (changes requested):
CI is red

Rohanraj123 commented:
Implemtation yet done : 
1. Used @ArrayLen() annotation at correct position as per the checker framework.
2. To process that annotation and verify . I added the command line option : `-Alint=trustArrayLenZero`
3. To maintain consistency among checkers profiles i added this command line arg.

Issue still : 
Nullchecker still doesnt process the annotation .
@romani

Rohanraj123 commented:
@romani Can you confirm me whether `Alint` command is configured correctly or not. 
Because if that is okay then something is wrong in  the CF tool. Because it doesnt recoganises the ArrayLen() annotation.

I have tried all the possible ways.

romani commented on pom.xml:
```diff
@@ -2536,6 +2536,8 @@
                 <Xmaxwarns>10000</Xmaxwarns>
               </compilerArguments>
+              <compilerArgument>-Xlint:all</compilerArgument>
+              <compilerArgument>-Alint=trustArrayLenZero </compilerArgument>
```
Try to put new lines before `</compilerArguments>`

Rohanraj123 replied:
done!

romani replied:
Not done

romani replied:
```
<compilerArguments>
                <Xlint>all</Xlint>
...
```

Rohanraj123 replied:
Nope It still aint work !

romani replied:
Let's ask for help from checkerframefork team, please create issue or discussion in their repository

Rohanraj123 replied:
Done : https://github.com/typetools/checker-framework/issues/6702

Rohanraj123 replied:
![Screenshot from 2024-07-09 06-49-34](https://github.com/typetools/checker-framework/assets/96484809/a1823c54-7c47-4cb3-89ed-3236f910d4db)

![Screenshot from 2024-07-09 06-49-31](https://github.com/typetools/checker-framework/assets/96484809/6bfc3520-00d4-427c-88b9-c86925a39eaf)

![Screenshot from 2024-07-09 06-49-16](https://github.com/typetools/checker-framework/assets/96484809/170eaaa3-5cad-494c-bf90-1d125b8c8d8f)

![Screenshot from 2024-07-09 06-49-09](https://github.com/typetools/checker-framework/assets/96484809/79320dc7-3c78-4f5f-88a8-9a2f59fd06f1)

@romani  These are the results i found when I actually configured checker framework in one of the demo project. 

It says Unsupported Command line option : Alint=trustArrayLenZero

romani replied:
Please share this test project in GitHub and show it to it maintainer, to let them suggest how to configure it.

romani replied:
There's also `<compilerArgs combine.children="append">` try to put it there

Rohanraj123 replied:
> There's also `<compilerArgs combine.children="append">` try to put it there

Its already there !

romani reviewed (changes requested):
Item

romani commented:
for now this PR is blocked by https://github.com/typetools/checker-framework/issues/6702

Rohanraj123 commented:
@romani This is the only way to place annotation I have tried all the ways in my demo project. 
But CI gives syntax error. Should I work on resolving the  syntax error or should we ditch the idea of using annotation?

romani commented:
Yes, please make code compile able.

Rohanraj123 commented:
@romani After looking into the problem. I found that plexus is generating this issue because not being able to generate the meta data due to syntax incompatibility. Now The only way is to change the place of annotation should not be there or excluding pexus from our codebase and using something else that can just be compatible with any syntax type. 

Do you have somework around in your mind that we can implement ? Please suggest!

Rohanraj123 commented:
One more last thing i wanna confirm from you. I am thinking of using this workaround that I preprocess  the files and temporarily remove the annotations during meta data generation and then again restore it afterwards. I can write scripts for this purpose and and integerate in pom.xml file.

romani commented:
No, add/Remove is super hack, looks like we need to go deeper in plugin that failing on such syntax.

romani commented:
we lost energy in this PR, I closing it, we will be back to it one day in future.
